### PR TITLE
fix: [CDS-49039]: Fixing the execution flow for openshift templates manifest

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Fixing the execution flow for openshift templates manifest using custom remote store. Kubernetes deployment was not working with Openshift using custom remote store.

1. No Params Git Store
Screenshot 2022-12-29 at 12 50 07 AM

2. Openshift Template (Git) + Openshift Params (Custom + Harness + Git)
Screenshot 2022-12-29 at 12 50 25 AM

3. Openshift Template (Harness) + Openshift Params (Custom + Harness + Git)
Screenshot 2022-12-29 at 12 51 48 AM

4. Openshift Template (Custom) + Openshift Params (Custom + Harness + Git)
Screenshot 2022-12-29 at 12 52 22 AM

5. Openshift Template (Custom)
Screenshot 2022-12-29 at 12 52 43 AM

6. Openshift Template (Custom) + Openshift Params (Custom)
Screenshot 2022-12-29 at 12 52 58 AM

## Automation 
Added UT
Automation will be a part of this ticket: https://harness.atlassian.net/browse/CDS-49064

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41875)

### CODE REVIEW

**Review by abosii-harness — CHANGES_REQUESTED** [review]
Please make sure that all scenarios are working properly. 
1. No Params
2. Git 
3. Git + Local 
4. Git + Local + Custom
5. Local
6. Local + Custom
7. Custom

The overall logic is very tricky and hard to follow

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java [hunk 1]
```diff
@@ -3133,4 +3121,25 @@ private static String getLatestVersionOcPath() {
     }
     return ocPath;
   }
+
+  private String getOpenshiftTemplatePath(StoreDelegateConfig storeDelegateConfig, ManifestType manifestType) {
+    if (storeDelegateConfig instanceof GitStoreDelegateConfig) {
+      GitStoreDelegateConfig otGitStoreDelegateConfig = (GitStoreDelegateConfig) storeDelegateConfig;
+      return otGitStoreDelegateConfig.getPaths().get(0);
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java:3128]: Can we make sure that the paths is not empty and if it is throw the proper error? This should be done for other branches, something like InvalidArgumentException should be enough, but if you can think about some hint + explanation that will be good

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java [hunk 2]
```diff
@@ -3133,4 +3121,25 @@ private static String getLatestVersionOcPath() {
...
+    } else if (storeDelegateConfig instanceof LocalFileStoreDelegateConfig) {
+      LocalFileStoreDelegateConfig localFileStoreDelegateConfig = (LocalFileStoreDelegateConfig) storeDelegateConfig;
+      String openshiftTemplatePath = localFileStoreDelegateConfig.getManifestFiles().get(0).getFilePath();
+      if (openshiftTemplatePath.charAt(0) == '/') {
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java:3132]: We should have an isNotEmpty(openshiftTemplatePath) check before openshiftTemplatePath.charAt(0)

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java [hunk 3]
```diff
@@ -3133,4 +3121,25 @@ private static String getLatestVersionOcPath() {
...
+        openshiftTemplatePath = openshiftTemplatePath.substring(1);
+      }
+      return openshiftTemplatePath;
+    } else if (storeDelegateConfig instanceof CustomRemoteStoreDelegateConfig) {
+      CustomRemoteStoreDelegateConfig customRemoteStoreDelegateConfig =
+          (CustomRemoteStoreDelegateConfig) storeDelegateConfig;
+      return getFileName(customRemoteStoreDelegateConfig.getCustomManifestSource().getFilePaths().get(0));
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java:3139]: Why this getFileName method is required here? Didn't see it before. Anyway is comming from fabric8 
import static io.fabric8.utils.Files.getFileName;
First we should avoid using fabric8 and the method logic is not defined by us

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sStepHelper.java [hunk 4]
```diff
@@ -742,8 +742,16 @@ private TaskChainResponse handleCustomFetchResponse(ResponseData responseData, K
...
       }
 
-      return executeOpenShiftParamsTask(
-          ambiance, stepElementParameters, openshiftParamManifestOutcomes, updatedK8sStepPassThroughData);
+      if (shouldExecuteGitFetchTask(openshiftParamManifestOutcomes)
+          || ManifestStoreType.isInGitSubset(k8sManifest.getStore().getKind())) {
+        return executeOpenShiftParamsTask(
+            ambiance, stepElementParameters, openshiftParamManifestOutcomes, updatedK8sStepPassThroughData);
+      } else {
+        LinkedList<ManifestOutcome> orderedParamsManifests = new LinkedList<>(openshiftParamManifestOutcomes);
```
Comment (abosii-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sStepHelper.java:750]: Is this conversion from ArrayList -> LinkedList -> ArrayList is really required?

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41875)* status: 6 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sStepHelper.java` line 750 at r3](https://reviewable.io/reviews/harness/harness-core/41875#-NJoXNBzn45PdMDa_YJd-r3-750:-NKOyWSeATnIRmTO7i7c:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/aea8c4dd9c01f415d7c5e8cad5b98512d1c0d8cb/125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sStepHelper.java#L750)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Is this conversion from ArrayList -> LinkedList -> ArrayList is really required?
</blockquote></details>

Done
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java` line 3128 at r3](https://reviewable.io/reviews/harness/harness-core/41875#-NJoXNBzn45PdMDa_YJf-r3-3128:-NKOy_E15Pzs3r6Nf9WD:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/aea8c4dd9c01f415d7c5e8cad5b98512d1c0d8cb/930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java#L3128)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Can we make sure that the paths is not empty and if it is throw the proper error? This should be done for other branches, something like InvalidArgumentException should be enough, but if you can think about some hint + explanation that will be good 
</blockquote></details>

Done
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java` line 3132 at r3](https://reviewable.io/reviews/harness/harness-core/41875#-NJoXNBzn45PdMDa_YJf-r3-3132:-NKOyaIv1BjBDiaFwTse:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/aea8c4dd9c01f415d7c5e8cad5b98512d1c0d8cb/930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java#L3132)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

We should have an isNotEmpty(openshiftTemplatePath) check before openshiftTemplatePath.charAt(0)
</blockquote></details>

Done
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java` line 3139 at r3](https://reviewable.io/reviews/harness/harness-core/41875#-NJoXNBzn45PdMDa_YJf-r3-3139:-NKOybKJ11BJ9qXCYH2m:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/aea8c4dd9c01f415d7c5e8cad5b98512d1c0d8cb/930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/K8sTaskHelperBase.java#L3139)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Why this getFileName method is required here? Didn't see it before. Anyway is comming from fabric8 
import static io.fabric8.utils.Files.getFileName;
First we should avoid using fabric8 and the method logic is not defined by us
</blockquote></details>

Done

**Review by abosii-harness — APPROVED** [review]
> Reviewed 4 of 6 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41875)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Review by achyuth-harness — APPROVED** [review]
> Reviewed 2 of 6 files at r4, 2 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41875)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by abosii-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41875)* status: :shipit: complete! ✓
